torch/csrc/jit/operator.cpp

```diff
@@ -62,25 +62,48 @@ struct SchemaParser {
       throw ErrorReport(tok.range) << "unknown type specifier";
     return it->second;
   }
-  void parseType(Argument& arg) {
-    arg.type = parseBaseType();
-    if(L.nextIf('[')) {
-      arg.type = ListType::create(arg.type);
-      if(L.cur().kind == TK_NUMBER) {
-        arg.N = std::stoll(L.next().text());
+  void parseType(std::vector<TypePtr>& types) {
+    TypePtr type;
+    if (L.cur().kind == '(') {
+      std::vector<TypePtr> nestedTypes;
+      parseList('(', ',', ')', nestedTypes, &SchemaParser::parseType);
+      type = TupleType::create(std::move(nestedTypes));
+    } else {
+      type = parseBaseType();
+      if(L.nextIf('[')) {
+        type = ListType::create(type);
+        if(L.cur().kind == TK_NUMBER) {
+          L.next(); // Discard
+        }
+        L.expect(']');
+      }
+    }
+    types.push_back(std::move(type));
+  }
+  void parseArgumentType(Argument& arg) {
```

Comment:
Let's unify these and change it to `parseArgumentType(TypePtr* type, size_t* N)`. Then you can implement `parseType(std::vector<TypePtr>& types) { types.push_back(); parseArgumentType(types.back(), nullptr); }`